rclcpp/include/rclcpp/node.hpp

```diff
@@ -258,133 +258,584 @@ class Node : public std::enable_shared_from_this<Node>
 
+  /// Return the parameters by the given parameter names.
+  /**
+   * Like get_parameters(), this method may throw the
+   * rclcpp::exceptions::ParameterNotDeclaredException exception if the
+   * requested parameter has not been declared and undeclared parameters are
+   * not allowed.
+   *
+   * Also like get_parameters(), if undeclared parameter are allowed and the
+   * parameter has not been declared, then the corresponding rclcpp::Parameter
+   * will be default initialized and therefore have the type
+   * rclcpp::ParameterType::PARAMETER_NOT_SET.
+   *
+   * \param[in] names The names of the parameters to be retrieved.
+   * \return The parameters that were retrieved.
+   * \throws rclcpp::exceptions::ParameterNotDeclaredException if any of the
+   *   parameters have not been declared and undeclared parameters are not
+   *   allowed.
+   */
+  RCLCPP_PUBLIC
+  std::vector<rclcpp::Parameter>
+  get_parameters(const std::vector<std::string> & names) const;
+
+  /// Get the parameter values for all parameters which are prefixed with the given prefix.
+  /**
+   * The "prefix" argument is used to list the parameters which are prefixed
+   * with that prefix, see also list_parameters().
+   *
+   * The resulting list of parameter names are used to get the values of the
+   * parameters.
+   *
+   * The names which are used as keys in the values map have the prefix removed.
+   * For example, if you use the prefix "foo" and the parameters "foo.ping" and
+   * "foo.pong" exist, then the returned map will have the keys "ping" and
+   * "pong".
+   *
+   * An empty string for the prefix will match all parameters.
+   *
+   * If no parameters with the prefix are found, then the output parameter
+   * "values" will be unchanged and false will be returned.
+   * Otherwise, the parameter names and values will be stored in the map and
+   * true will be returned to indicate "values" was mutated.
+   *
+   * This method will never throw the
+   * rclcpp::exceptions::ParameterNotDeclaredException exception because the
+   * action of listing the parameters is done atomically with getting the
+   * values, and therefore they are only listed if already declared and cannot
+   * be undeclared before being retrieved.
+   *
+   * Like the templated get_parameter() variant, this method will attempt to
+   * coerce the parameter values into the type requested by the given
+   * template argument, which may fail and throw an exception.
+   *
+   * \param[in] prefix The prefix of the parameters to get.
+   * \param[out] values The map used to store the parameter names and values,
+   *   respectively, with one entry per parameter matching prefix.
+   * \returns True if output "values" was changed, false otherwise.
+   * \throws rclcpp::ParameterTypeException if the requested type does not
+   *   match the value of the parameter which is stored.
+   */
+  template<typename ParameterT>
+  bool
+  get_parameters(
+    const std::string & prefix,
```

Comment:
Yeah, this is a bit wonky, because you can do something like `foo.b` to match `foo.bar` and `foo.baz`... The addressing of parameters needs to be cleaned up further and the terminology needs to be clearer, but I didn't have time to follow through on that.

Reply:
Sorry, didn't complete my thought... So that means in this case it's a prefix rather than a namespace technically, though a namespace is also technically a kind of prefix, and this function will remove any leading namespace separator as I described in the docs. But at this point I'm just trying to document existing behavior, I don't think this is the best way to have all this work...